src/snowflake/connector/connection.py

```diff
@@ -1086,16 +1090,29 @@ def _process_params_qmarks(
                             "errno": ER_NOT_IMPLICITY_SNOWFLAKE_DATATYPE,
                         },
                     )
-                if isinstance(v, list):
-                    vv = [
-                        self.converter.to_snowflake_bindings(
-                            self.converter.snowflake_type(v0), v0
-                        )
-                        for v0 in v
-                    ]
-                else:
-                    vv = self.converter.to_snowflake_bindings(snowflake_type, v)
-                processed_params[str(idx + 1)] = {"type": snowflake_type, "value": vv}
+                return TypeAndBinding(
+                    snowflake_type,
+                    self.converter.to_snowflake_bindings(snowflake_type, v),
+                )
+
+        for idx, v in enumerate(params):
+            if isinstance(v, list):
+                snowflake_type = self.converter.snowflake_type(v)
+                all_param_data = list(map(get_snowflake_type_and_binding, v))
+                first_type = all_param_data[0].type
+                # if all elements have the same snowflake type, update snowflake_type
+                if all(param_data.type == first_type for param_data in all_param_data):
+                    snowflake_type = first_type
```

Comment:
This is to set `snowflake_type` to "LIST", then it obtains the type and binding for each parameter in the list.
If all types are the same -> update snowflake type.